server/controllers/finance/reports/income_expense_by_year/index.js

```diff
@@ -0,0 +1,308 @@
+        balance : totals.income.balance + totals.expense.balance,
+        thirdBalance : totals.income.thirdBalance + totals.expense.thirdBalance,
+      };
+
+      _.extend(data, {
+        profits,
+        losses,
+        totals,
+      });
+
+      return report.render(data);
+    })
+    .then(result => {
+      res.set(result.headers).send(result.report);
+    })
+    .catch(next);
+}
+
+/**
+ * @function getAccountBalances
+ *
+ */
+function getAccountBalances(fiscalYearId) {
+  const sql = `
+    SELECT a.id, a.number, a.label, a.type_id, a.label, a.parent,
+      a.type_id = ${TITLE_ID} AS isTitleAccount, balance
+    FROM account AS a LEFT JOIN (
+      SELECT pt.account_id, SUM(pt.debit - pt.credit) AS balance
+      FROM period_total AS pt
+      JOIN period AS p ON p.id = pt.period_id
+      WHERE pt.fiscal_year_id = ?
+      GROUP BY pt.account_id
+    )s ON a.id = s.account_id
+    WHERE a.type_id IN (?)
+    ORDER BY a.number;
+  `;
+
+  const accountTypes = [
+    TITLE_ID,
+    INCOME_TYPE_ID,
+    EXPENSE_TYPE_ID,
+  ];
+
+  const params = [fiscalYearId, accountTypes];
+  return db.exec(sql, params);
+}
+
+/**
+ * @function combineIntoSingleDataset
+ *
+ * @description
+ * This function takes two partially overlapping datasets and combines them into
+ * a single dataset.  The current balances set the balance values and the
+ * previous balances are assigned to firstBalance property.  Missing data is
+ * filled in with zeroes.
+ *
+ * @returns Array
+ */
+function combineIntoSingleDataset(secondBalances, firstBalances, thirdBalances) {
+  // make an id -> account map for both values
+  const secondMap = new Map(secondBalances.map(a => [a.id, a]));
+  const firstMap = new Map(firstBalances.map(a => [a.id, a]));
+  const thirdMap = new Map(thirdBalances.map(a => [a.id, a]));
+
+  // get all account ids in a single list.
+  const combined = _.uniq([...secondMap.keys(), ...firstMap.keys(), ...thirdMap.keys()]);
+
+  return combined.map(id => {
+    const record = {};
+
+    const second = secondMap.get(id);
+    const first = firstMap.get(id);
+    const third = thirdMap.get(id);
+
+    if (second) {
+      const firstBalance = first ? first.balance : 0;
+      const thirdBalance = third ? third.balance : 0;
+      _.extend(record, second, {
+        firstBalance,
+        thirdBalance,
+      });
+    } else if (first) {
+      const thirdBalance = third ? third.balance : 0;
+      _.extend(record, first, {
+        balance : 0,
+      }, {
+        firstBalance : first.balance,
+        thirdBalance,
+      });
+    } else {
+      _.extend(record, third, {
+        balance : 0,
+        firstBalance : 0,
+      }, {
+        thirdBalance : third.balance,
+      });
+    }
+
+    return record;
+  });
+}
+
+/**
+ * @function getPeriodByNumberAndFiscalId
+ *
+ * @description
+ * Small helper to get all properties of a period by its number.
+ */
+function getFiscalById(fiscalId) {
```

Comment:
Can you just use `Fiscal.lookupFiscalYear(id)`?  We want to repeat ourselves as few times as possible.